src/Common/src/TypeSystem/IL/Stubs/EnumThunks.cs

```diff
@@ -0,0 +1,219 @@
+            }
+        }
+
+        public override TypeSystemContext Context
+        {
+            get
+            {
+                return _owningType.Context;
+            }
+        }
+
+        public override TypeDesc OwningType
+        {
+            get
+            {
+                return _owningType;
+            }
+        }
+
+        public override MethodSignature Signature
+        {
+            get
+            {
+                return _signature;
+            }
+        }
+
+        public override string Name
+        {
+            get
+            {
+                return "Equals";
+            }
+        }
+
+        public override bool IsVirtual
+        {
+            get
+            {
+                // This would be implicit (false is the default), but I wanted to make it explicit.
+                // Making this an actual override would cause size bloat with very little benefit.
+                // The usefulness of this method lies in it's ability to prevent boxing of 'this'.
+                // The base implementation on System.Enum is adequate for everything else.
+                return false;
+            }
+        }
+
+        public override MethodIL EmitIL()
+        {
+            ILEmitter emitter = new ILEmitter();
+            ILCodeStream codeStream = emitter.NewCodeStream();
+
+            // InstantiateAsOpen covers the weird case of generic enums
+            MetadataType owningTypeAsOpen = ((MetadataType)_owningType).InstantiateAsOpen();
+
+            ILCodeLabel lNotEqual = emitter.NewCodeLabel();
+
+            // if (!(obj is {enumtype}))
+            //     return false;
+
+            codeStream.EmitLdArg(1);
+            codeStream.Emit(ILOpcode.isinst, emitter.NewToken(owningTypeAsOpen));
+            codeStream.Emit(ILOpcode.dup);
+            codeStream.Emit(ILOpcode.brfalse, lNotEqual);
+
+            // return ({underlyingtype})this == ({underlyingtype})obj;
+
+            ILOpcode loadIndirectOpcode;
+            switch (_owningType.GetElementSize())
+            {
+                case 1:
+                    loadIndirectOpcode = ILOpcode.ldind_i1; break;
+                case 2:
+                    loadIndirectOpcode = ILOpcode.ldind_i2; break;
+                case 4:
+                    loadIndirectOpcode = ILOpcode.ldind_i4; break;
+                case 8:
+                    loadIndirectOpcode = ILOpcode.ldind_i8; break;
+                default:
+                    throw new InvalidOperationException();
+            }
+
+            // PREFER: ILOpcode.unbox, but the codegen for that is pretty bad
+            codeStream.Emit(ILOpcode.ldflda, emitter.NewToken(Context.GetWellKnownType(WellKnownType.Object).GetKnownField("m_pEEType")));
+            codeStream.EmitLdc(Context.Target.PointerSize);
+            codeStream.Emit(ILOpcode.add);
+            codeStream.Emit(loadIndirectOpcode);
```

Comment:
Note to self: can be replaced with `ldobj` to get rid of the 10 line switch block...